src/main/java/carpet/network/ClientNetworkHandler.java

```diff
@@ -19,36 +21,58 @@
 
 public class ClientNetworkHandler
 {
-    private static Map<String, BiConsumer<ClientPlayerEntity, Tag>> dataHandlers = new HashMap<String, BiConsumer<ClientPlayerEntity, Tag>>(){{
-        put("Rules", (p, t) -> {
+    private static Map<String, BiConsumer<ClientPlayerEntity, Tag>> dataHandlers = new HashMap<String, BiConsumer<ClientPlayerEntity, Tag>>();
+    static
+    {
+        dataHandlers.put("Rules", (p, t) -> {
             CompoundTag ruleset = (CompoundTag)t;
-            for (String ruleName: ruleset.getKeys())
+            for (String fullRuleName: ruleset.getKeys())
             {
-                ParsedRule<?> rule = CarpetServer.settingsManager.getRule(ruleName);
+                int separatorPos = fullRuleName.indexOf(':');
+                SettingsManager manager = null;
+                String ruleName;
+                if (separatorPos > -1)
+                {
+                    String identifier = fullRuleName.substring(0, separatorPos);
+                    ruleName = fullRuleName.substring(separatorPos + 1);
+                    for (CarpetExtension extension: CarpetServer.extensions) {
+                        SettingsManager eManager = extension.customSettingsManager();
+                        if (eManager != null && identifier.equals(eManager.getIdentifier()))
+                        {
+                            manager = eManager;
+                            break;
+                        }
+                    }
+                }
+                else
+                {
+                    ruleName = fullRuleName;
+                    manager = CarpetServer.settingsManager;
+                }
+                ParsedRule<?> rule = (manager != null) ? manager.getRule(ruleName) : null;
                 if (rule == null)
-                    CarpetSettings.LOG.error("Received unknown rule: "+ruleName);
+                    CarpetSettings.LOG.error("Received unknown rule: " + fullRuleName);
```

Comment:
I think that should be quiet - when joining carpet with extension you don't have.

Syncing extensions is yet another topic. Can be discussed later.

Reply:
Ok, removing. I just kept it because it was already there.